cbindgen/cbindgen.bzl

```diff
@@ -0,0 +1,220 @@
+            output_header.short_path,
+        ),
+        outputs = [output_header],
+        executable = cbindgen_bin,
+        inputs = rust_lib[CrateInfo].srcs + [ctx.outputs.config],
+        arguments = [args],
+    )
+
+    rust_compilation_context = rust_lib[CcInfo].compilation_context
+
+    # Add the new headers to the existing CompilationContext info
+    compilation_context = cc_common.create_compilation_context(
+        headers = depset([output_header], transitive = [rust_compilation_context.headers]),
+        defines = rust_compilation_context.defines,
+        framework_includes = rust_compilation_context.framework_includes,
+        includes = rust_compilation_context.includes,
+        local_defines = rust_compilation_context.local_defines,
+        quote_includes = rust_compilation_context.quote_includes,
+        system_includes = rust_compilation_context.system_includes,
+    )
+
+    # Return all providers given by `rust_library` to ensure compatiblity with other rules
+    return [
+        rust_lib[CrateInfo],
+        rust_lib[DepInfo],
+        CcInfo(
+            compilation_context = compilation_context,
+            linking_context = rust_lib[CcInfo].linking_context,
+        ),
+        DefaultInfo(
+            files = depset([output_header], transitive = [rust_lib.files]),
+            runfiles = ctx.runfiles([output_header], transitive_files = rust_lib.files),
+        ),
+    ]
+
+rust_cbindgen = rule(
+    implementation = _rust_cbindgen_impl,
+    doc = _rust_cbindgen_doc,
+    attrs = {
+        "lib": attr.label(
+            doc = (
+                "The `rust_library` target from which to run cbindgen on. " +
+                "The `create_type` of the target passed here must be " +
+                "either `cdylib` or `staticlib`."
+            ),
+            providers = [CrateInfo, CcInfo],
+            mandatory = True,
+        ),
+        "cbindgen_flags": attr.string_list(
+            doc = (
+                "Optional flags to pass directly to the bindgen executable. " +
+                "See https://docs.rs/cbindgen/latest/cbindgen/ for details."
+            ),
+        ),
+        "header_name": attr.string(
+            doc = (
+                "Optional override for the name of the generated header. The default is the " +
+                "name of the target created by this rule."
+            ),
+        ),
+        "lang": attr.string(
+            doc = "Optional target language identifier of the generated header file",
+            values = [
+                "c",
+                "c++",
+            ],
+            default = "c++",
+        ),
+        "config": attr.label(
+            doc = "Optional Cbindgen configuration template",
+            allow_single_file = True,
+        ),
+        "substitutions": attr.string_dict(
+            doc = "Optional substitutions for the cbindgen config template passed to 'config'",
+        ),
+        "cbindgen": attr.label(
```

Comment:
I think it's fine to provide all these ways to pass in cbindgen, but "The Right Way" is probably to use a toolchain

Reply:
I'm not too familiar with toolchains. I'm happy to take a crack at it though. Any suggestions on how to start?

Reply:
Are you suggesting I add to that toolchain or make a new one similar to it?

Reply:
I made a toolchain for it